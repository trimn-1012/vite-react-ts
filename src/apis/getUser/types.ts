import { IUser } from '../getUsers/types';

export type IUserResponse = IUser;
