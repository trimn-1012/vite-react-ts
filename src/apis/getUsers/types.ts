import { IUser } from '@/apis/getUser/types';

export interface IUsersResponse {
  users: IUser[];
}
