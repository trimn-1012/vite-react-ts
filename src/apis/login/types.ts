import type { Merge } from 'type-fest';

import { IUser } from '@/apis/getUser/types';

interface IToken {
  token: string;
}

export type TLoginParams = {
  username: string;
  password: string;
};

export type ILoginResponse = Merge<IUser, IToken>;
