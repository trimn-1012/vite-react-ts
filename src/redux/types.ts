import store from './store';

export type IRootState = ReturnType<typeof store.getState>;
