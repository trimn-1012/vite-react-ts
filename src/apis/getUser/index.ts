import request from '@/services/request';

import { IUserResponse } from './types';

const keyUser = '/user';

async function fetchUser(url: string) {
  const res = await request<IUserResponse>({
    method: 'GET',
    url,
  });

  return res.data;
}

export { keyUser, fetchUser };
