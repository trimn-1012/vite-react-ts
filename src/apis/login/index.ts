import { ILoginResponse, LoginParams } from './types';
import request from '../../services/request';

const keyLogin = '/auth/login';

async function login(url: string, { arg }: { arg: LoginParams }) {
  const res = await request<ILoginResponse>({
    method: 'POST',
    url,
    data: {
      username: arg.username,
      password: arg.password,
    },
  });

  return res.data;
}

export { keyLogin, login };
