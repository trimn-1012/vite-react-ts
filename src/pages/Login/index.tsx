import { Controller, useForm, SubmitHandler } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { yupResolver } from '@hookform/resolvers/yup';

import yup from '@/utils/yupGlobal';
import routes from '@/routes/paths';
import tokenStorage from '@/utils/tokenStorage';
import { useLoginMutation } from '@/hooks/useLoginMutation';

type FormValues = {
  username: string;
  password: string;
};

const schema: yup.ObjectSchema<FormValues> = yup.object().shape({
  username: yup.string().required('username is required'),
  password: yup.string().required('password is required'),
});

const Login = () => {
  const navigate = useNavigate();
  const { handleSubmit, control, formState } = useForm<FormValues>({
    mode: 'onChange',
    defaultValues: {
      username: '',
      password: '',
    },
    resolver: yupResolver(schema),
  });

  const { trigger: loginMutation, isMutating } = useLoginMutation();

  const onSubmit: SubmitHandler<FormValues> = async data => {
    await loginMutation(data, {
      onSuccess: successData => {
        alert('Login successfully');
        tokenStorage.set(successData.id.toString());
        navigate(routes.users.build());
      },
      onError: error => {
        if (error.data) alert(error.data.message);
      },
    });
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      Login
      <div>
        <label htmlFor="username">Username</label>
        <Controller
          control={control}
          name="username"
          render={({ field }) => (
            <input
              data-testid="username"
              id="username"
              type="text"
              {...field}
            />
          )}
        />
        {!!formState.errors.username && (
          <p>{formState.errors.username.message}</p>
        )}
      </div>
      <div>
        <label htmlFor="password">Password</label>
        <Controller
          control={control}
          name="password"
          render={({ field }) => (
            <input
              data-testid="password"
              id="password"
              type="password"
              {...field}
            />
          )}
        />
        {!!formState.errors.password && (
          <p>{formState.errors.password.message}</p>
        )}
      </div>
      <button data-testid="loginButton" disabled={isMutating}>
        Login
      </button>
    </form>
  );
};

export default Login;
