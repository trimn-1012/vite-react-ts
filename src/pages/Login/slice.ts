import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { IUser } from '@/apis/getUser/types';
import { IRootState } from '@/redux/types';

interface ISliceState {
  user: IUser | null;
}

const initialState: ISliceState = {
  user: null,
};
const slice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    updateUser: (state, action: PayloadAction<IUser>) => {
      state.user = action.payload;
    },
  },
});

const { reducer, actions } = slice;

export const { updateUser } = actions;

export const userSelector = (state: IRootState) => state.login.user;
export const isAuthenticatedSelector = (state: IRootState) =>
  !!state.login.user;

export default reducer;
