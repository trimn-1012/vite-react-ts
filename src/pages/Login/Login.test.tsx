import { configureStore, Store } from '@reduxjs/toolkit';
import { fireEvent, waitFor } from '@testing-library/react';
import { Mock, vi } from 'vitest';
import MockAdapter from 'axios-mock-adapter';

import { userEvent, renderWithAllProviders } from '@/utils/testUtils';
import { IUser } from '@/apis/getUser/types';
import { IRootState } from '@/redux/types';
import appAxios from '@/services/appAxios';
import { keyLogin } from '@/apis/login';

import Login from './index';
import reducer, {
  initialState,
  updateUser,
  userSelector,
  isAuthenticatedSelector,
} from './slice';

let store: Store<Pick<IRootState, 'login'>>;

const mockData: IUser = {
  id: 1,
  firstName: 'Terry',
  lastName: 'Medhurst',
  maidenName: 'Smitham',
  age: 50,
  gender: 'male',
  email: 'atuny0@sohu.com',
  phone: '+63 791 675 8914',
  username: 'atuny0',
  password: '9uQFF1Lh',
  birthDate: '2000-12-25',
  image: 'https://robohash.org/hicveldicta.png',
  bloodGroup: 'Aâˆ’',
};

describe('pages/Login', () => {
  describe('slice.ts', () => {
    beforeEach(() => {
      store = configureStore({
        reducer: {
          login: reducer,
        },
      });
    });

    it('should return the initial state', () => {
      const state = store.getState();

      expect(state.login).toEqual(initialState);
    });

    it('should handle updateUser action', () => {
      store.dispatch(updateUser(mockData));
      const state = store.getState();

      expect(userSelector(state)).toEqual(mockData);
    });

    it('should isAuthenticated return true if user exist', () => {
      store.dispatch(updateUser(mockData));
      const state = store.getState();

      expect(isAuthenticatedSelector(state)).toEqual(true);
    });

    it('should isAuthenticated return false if user null', () => {
      const state = store.getState();

      expect(isAuthenticatedSelector(state)).toEqual(false);
    });
  });

  describe('index.tsx', () => {
    describe('snapshot', () => {
      it('should render and match the snapshot', () => {
        const { container } = renderWithAllProviders(<Login />);

        expect(container.firstChild).toMatchSnapshot();
      });
    });
    describe('function', () => {
      let mockedAlert: Mock<any, any>;
      let mockAxios: MockAdapter;

      beforeEach(() => {
        mockAxios = new MockAdapter(appAxios);
        mockedAlert = vi.fn();
        global.alert = mockedAlert;
      });

      it('renders login form with username, password input and login button', () => {
        const { getByTestId } = renderWithAllProviders(<Login />);
        const usernameElement = getByTestId('username');
        const passwordElement = getByTestId('password');
        const buttonElement = getByTestId('loginButton');
        expect(usernameElement).toBeInTheDocument();
        expect(passwordElement).toBeInTheDocument();
        expect(buttonElement).toBeInTheDocument();
      });

      it('should update state when input values change', () => {
        const { getByTestId } = renderWithAllProviders(<Login />);
        const usernameInput = getByTestId('username');
        const passwordInput = getByTestId('password');

        fireEvent.change(usernameInput, { target: { value: 'username' } });
        fireEvent.change(passwordInput, { target: { value: 'password' } });

        expect(usernameInput).toHaveValue('username');
        expect(passwordInput).toHaveValue('password');
      });

      it('calls login API with correct username and password', async () => {
        const user = userEvent.setup();
        const { getByTestId, getByText } = renderWithAllProviders(<Login />);
        const usernameElement = getByTestId('username');
        const passwordElement = getByTestId('password');
        const buttonElement = getByTestId('loginButton');

        // Click the submit button without filling the form
        await user.click(buttonElement);

        expect(getByText('username is required')).toBeInTheDocument();
        expect(getByText('password is required')).toBeInTheDocument();

        fireEvent.change(usernameElement, { target: { value: 'atuny0' } });
        fireEvent.change(passwordElement, { target: { value: '9uQFF1Lh' } });

        mockAxios
          .onPost(keyLogin, {
            username: 'atuny0',
            password: '9uQFF1Lh',
          })
          .reply(200, mockData);

        fireEvent.click(buttonElement);

        await waitFor(() => {
          expect(mockedAlert).toHaveBeenCalledWith('Login successfully');
        });
      });
    });
  });
});
