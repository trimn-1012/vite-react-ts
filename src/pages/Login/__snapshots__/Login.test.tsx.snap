// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pages/Login > index.tsx > snapshot > should render and match the snapshot 1`] = `
<form>
  Login
  <div>
    <label
      for="username"
    >
      Username
    </label>
    <input
      data-testid="username"
      id="username"
      name="username"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      for="password"
    >
      Password
    </label>
    <input
      data-testid="password"
      id="password"
      name="password"
      type="password"
      value=""
    />
  </div>
  <button
    data-testid="loginButton"
  >
    Login
  </button>
</form>
`;
